core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -0,0 +1,378 @@
+package cats
+package data
+
+import NonEmptyLazyList.create
+import kernel.PartialOrder
+import instances.lazyList._
+
+import scala.collection.immutable.TreeSet
+
+object NonEmptyLazyList extends NonEmptyLazyListInstances {
+  private[data] type Base
+  private[data] trait Tag extends Any
+  type Type[+A] <: Base with Tag
```

Comment:
I don’t understand how these are abstract on a concrete `object`. What is Type?

Reply:
This is the newtype encoding (kind of a hack of the Scala type system) that eliminates all runtime overhead.

Reply:
I know the trick, I just didn't know objects could have totally abstract types. Is that intended, or some quirk of the compiler?

Seems to me, anything we an abstract type should have to be abstract.

Do we have this documented somewhere that we can link to? I think it is not at all clear to the reader what the various parts are here and why they are all required:

1. why do we need Base?
2. why do we need Tag?
3. why should the above be `private[data]` vs `private` or `public`?
4. how is the aliasing of this type to `NonEmptyLazyList` accomplished? Is that a package alias?

Reply:
good point. Link added.

Reply:
also cc @LukaJCB to see if he has anything to add to the simple description and the link